### PR TITLE
error handling proposal

### LINKED ISSUE #141: Should span creation operations raise exceptions?
From https://github.com/open-telemetry/opentelemetry-python/pull/11#pullrequestreview-245809272.

The java tracer API docs note that calling `spanBuilder` with a null `spanName` arg should [raise a `NullPointerException`](https://github.com/open-telemetry/opentelemetry-java/blob/8520de558133c2e09f9f0de8b07f36c377b3555d/api/src/main/java/io/opentelemetry/trace/Tracer.java#L149). `NoopSpanBuilder` [does this](https://github.com/open-telemetry/opentelemetry-java/blob/8520de558133c2e09f9f0de8b07f36c377b3555d/api/src/main/java/io/opentelemetry/trace/DefaultTracer.java#L152-L153), but `SpanBuilderSdk` doesn't.

Two questions here:

Should we create a span with a default name (as @reyang and @carlosalberto suggest) instead of raising if the span name is null here? If we want to avoid exceptions in the API altogether this suggests we should also create a blank span current if the user calls `Tracer.withSpan` with a null `span`, but this seems less defensible than using a default name.

More generally, which errors should implementations raise? I could imagine letting the implementation be either more or less restrictive than the API, but if it can both raise exceptions that aren't listed in the API and choose _not_ to raise exceptions that are... then there doesn't seem to be much point to listing exceptions in the API.

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Fix #141

This proposal suggests the change comparing to the current API/SDK implementation in Java.

### CODE REVIEW

File: specification/error-handling.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Error handling in OpenTelemetry
```
Comment (songy23) [inline specification/error-handling.md]: You can also mention SDK may provide essential self-monitoring, such as tracing its own methods and setting error status there. As an example: https://github.com/census-instrumentation/opencensus-java/blob/2e218f1b2fda8847654ca4dbbcbf8b680775bedc/exporters/metrics/util/src/main/java/io/opencensus/exporter/metrics/util/MetricReader.java#L155-L175.

File: specification/error-handling.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+This makes some unique requirements for OpenTelemetry error handling practices.
+
+## Basic error handling principles
+
+OpenTelemetry SDK must not throw or leak unhandled or user unhandled exceptions.
+
+1. APIs must not throw unhandled exceptions when the API is used incorrectly by
+   the end user. Smart defaults should be used so that the SDK generally works.
+   For instance, name like `empty` MUST be used when `null` value was passed as
```
Comment (songy23) [inline specification/error-handling.md]: Hmmm I think it's good practice to validate the required argument when end users call our API and throw exceptions when it violates certain constraints. Consider other cases like name being too long, I think it's better to throw exceptions than silently truncate it or use defaults.

File: specification/error-handling.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+potentially only enabled in runtime. Which makes the use of SDK non-obvious for
+the end user and sometimes even outside of the application developer control.
+
+This makes some unique requirements for OpenTelemetry error handling practices.
+
+## Basic error handling principles
+
+OpenTelemetry SDK must not throw or leak unhandled or user unhandled exceptions.
+
+1. APIs must not throw unhandled exceptions when the API is used incorrectly by
```
Comment (bogdandrutu) [inline specification/error-handling.md]: This is an interesting point. I think I have a different opinion:
1. When a specific implementation is used no different checks are applied - means that if the API says a span name is valid if not null, then implementation cannot throw exception if empty.
1. API can throw exception for things like null span name because that is documented.

Happy to be convinced otherwise, but my understanding is that as long as we apply the same checks across the API and SDK we can throw exception for obvious things.

Comment (SergeyKanzhelev) [reply specification/error-handling.md]: one problem that you may run into is that arguments you are passing to telemetry API may be received from external source. So you have a `null` string. If you are smart and read the doc - you will check for `null` and pass some random name to the API. If you are smart and haven't read the doc - API will crash and potentially bring this request processing with you.

Both outcomes can be worked around for by applying the same smart default as customer would need to anyway.

File: specification/error-handling.md (hunk 2 above)
Comment (SergeyKanzhelev) [reply specification/error-handling.md]: Would it be better if application doesn't start due to some environment variable was misconfigured (for resource API for instance) or if application sent a slightly inconsistent telemetry?

Same for span name. Would you rather return `500` to customer or use `empty` for the `null` string? (see also comment for @bogdandrutu 's question

File: specification/error-handling.md (hunk 1 above)
Comment (SergeyKanzhelev) [reply specification/error-handling.md]: added a section

File: specification/error-handling.md (hunk 2 above)
Comment (reyang) [reply specification/error-handling.md]: We had some conversations in Python SDK https://github.com/open-telemetry/opentelemetry-python/pull/11#discussion_r290582148.

One thing to consider: there is no simple way to determine if span name is too long since different exporters/backends might have different limitations. We don't want the users to see the application running fine for exporter A, and start to see unhandled exceptions when they switch to exporter B.

File: specification/error-handling.md (hunk 3 above)
Comment (bogdandrutu) [reply specification/error-handling.md]: Sure, but running a "blind" process with no monitor at all because we cannot export data to the metrics backend, is that better? I feel that fail fast approach should be used here and crashing during the initialization is very good in this case.

Comment (reyang) [reply specification/error-handling.md]: I agree with @bogdandrutu on "API can throw exception for things like null span name because that is documented".

Comment (carlosalberto) [reply specification/error-handling.md]: > Both outcomes can be worked around for by applying the same smart default as customer would need to anyway.

I agree with this. At the same time, we need a balance for reporting errors for truly fatal things (i.e. " cannot export data to the metrics backend"), but otherwise throwing exceptions should stay at a minimum level. 

Specific case: I do think using a null name for `Span` could fallback to using a default name, instead of throwing. Documenting it is nice, but providing defaults for these kind of things is a good alternative ;)

File: specification/error-handling.md (hunk 2 above)
Comment (carlosalberto) [reply specification/error-handling.md]: > We don't want the users to see the application running fine for exporter A, and start to see unhandled exceptions when they switch to exporter B.

+1

File: specification/error-handling.md (hunk 3 above)
Comment (iredelmeier) [reply specification/error-handling.md]: I agree re: failing fast at initialization/construction time, but not in ways that might crash the program (or thread or ____) at runtime.

File: specification/error-handling.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+2. Every background operation callback, Task or Thread method should have a
+   global error handling set up (like `try{}catch` statement) to ensure
+   that exception from this asynchronous operation will not affect end-user app.
+3. Error handling in other cases MUST follow standard language practice. Which
+   is typically - reduce the scope of the error handler and add special
+   processing for the expected errors.
+4. Beware of any call to external callbacks or override-able interface. Expect
+   them to throw.
+
+## SDK self-diagnostics
```
Comment (iredelmeier) [inline specification/error-handling.md]: While very valuable, self-diagnostics are inherently complicated. This feels like something that needs a larger discussion.

Comment (jmacd) [inline specification/error-handling.md]: I believe that user-facing instrumentation APIs should never return errors (or throw them).
One reason that users should never receive these, is that the'll be tempted to turn around and call the instrumentation library with the error. 
The SDK will, however, encounter errors, and it a user has a right to know, but _not with in-line code_. In the opentracing Go library we addressed this by letting the application register a callback for receiving self diagnostics, including errors, that would allow the application to handle self diagnostics in a separate module.

Comment (arminru) [reply specification/error-handling.md]: We are doing exactly the same at Dynatrace with our agent SDK:
The SDK API never throws, since we do not want to change the application's behavior and risk crashing it if they don't catch everything properly, just because monitoring wouldn't work. In order to not let our SDK users in the dark and give them guidance for troubleshooting we also offer a diagnostic callback. It provides immediate logging output in case of errors (see https://github.com/Dynatrace/OneAgent-SDK#logging-callback). This turned out to work well and be pretty helpful in the past.

**Comment by tedsuo** [comment]
This has gone a bit stale. I believe @SergeyKanzhelev is on paternity leave. Any opinions on how we resolve this, or otherwise move it forwards?

**Comment by c24t** [comment]
@tedsuo I updated the PR to address the comments in a way that I hope is consistent with @SergeyKanzhelev's original proposal and his responses on this PR.

File: specification/error-handling.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+In most cases, the work that the library performs is not essential from the perspective of application business logic.
+We assume that users would prefer to lose telemetry data rather than have the library significantly change the behavior of the instrumented application.
+
+OpenTelemetry may be enabled via platform extensibility mechanisms, or dynamically loaded at runtime.
+This makes the use of the library non-obvious for end users, and may even be outside of the application developer's control.
+This makes for some unique requirements with respect to error handling.
+
+## Basic error handling principles
+
+OpenTelemetry implementations MUST NOT throw unhandled exceptions at run time.
```
Comment (c24t) [inline specification/error-handling.md]: I originally changed this to read:

> The OpenTelemetry API package MUST NOT expose (i.e. throw, return, or otherwise leak) unhandled exceptions to end users at run time.
> OpenTelemetry implementations MUST NOT expose exceptions that are not documented in the API.

to address comments at https://github.com/open-telemetry/opentelemetry-specification/pull/153/commits/1b6737a9343ef7e184254b6597a9a2f580e34ce3#r296402616, but on reflection I think it is actually better not to have any checked exceptions in the API. It's also more consistent with @SergeyKanzhelev's original text.

The big unresolved problem here is NPEs.

File: specification/error-handling.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+1. API methods MUST NOT throw unhandled exceptions when used incorrectly by end users.
```
Comment (c24t) [inline specification/error-handling.md]: "Throw unhandled exceptions" is kind of redundant, but may help distinguish between checked exceptions listed in docs/signatures and unchecked. Since I'm proposing outlawing checked exceptions too this might not be necessary.

File: specification/error-handling.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+   The API and SDK SHOULD provide safe defaults for missing or invalid arguments.
+   For instance, a name like `empty` may be used if the user passes in `null` as the span name argument during `Span` construction.
+2. The API or SDK may _fail fast_ and cause the application to fail on initialization, e.g. because of bad user config or a environment, but MUST NOT cause the application to fail at run time, e.g. due to dynamic config settings received from the Agent.
```
Comment (c24t) [inline specification/error-handling.md]: New line, added for @iredelmeier's comment at https://github.com/open-telemetry/opentelemetry-specification/pull/153/commits/1b6737a9343ef7e184254b6597a9a2f580e34ce3#r296935968.

File: specification/error-handling.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+3. The SDK MUST NOT throw unhandled exceptions for errors in their own operations.
+   For example, an exporter should not throw an exception when it cannot reach the endpoint to which it sends telemetry data.
```
Comment (c24t) [inline specification/error-handling.md]: I removed 

> SDK MUST NOT crash process by throwing exception or causing `OutOfMemoryException` when telemetry receiving endpoint cannot be reached

because performance guarantees seem out of scope for this doc.

File: specification/error-handling.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+## Guidance
+
+1. API methods that accept external callbacks MUST handle all errors.
+2. Background tasks (e.g. threads, asynchronous tasks, and spawned processes) should run in the context of a global error handler to ensure that exceptions do not affect the end user application.
+3. Long-running background tasks should not fail permanently in response to internal errors.
```
Comment (c24t) [inline specification/error-handling.md]: New line to expand on the point above.

File: specification/error-handling.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+5. Beware external callbacks and overrideable interfaces: Expect them to throw.
+
+## Self-diagnostics
+
+All OpenTelemetry libraries -- the API, SDK, exporters, instrumentation adapters, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
+
+One good example of such telemetry is a `Span` exporter that indicates how much time exporters spend uploading telemetry.
+Another example may be a metric exposed by a `SpanProcessor` that describes the current queue size of telemetry data to be uploaded.
+
+Whenever the library suppresses an error that would otherwise have been exposed to the user, the library SHOULD log the error using language-specific conventions.
```
Comment (c24t) [inline specification/error-handling.md]: This is a new line, I thought logging was worth mentioning separately from callbacks.

File: specification/error-handling.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+SDKs MAY expose callbacks to allow end users to handle self-diagnostics separately from application code.
```
Comment (c24t) [inline specification/error-handling.md]: New line for @jmacd and @arminru's comments at https://github.com/open-telemetry/opentelemetry-specification/pull/153#discussion_r297810229.

File: specification/error-handling.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+
+## Exceptions to the rule
+
+SDK authors MAY supply settings that allow end users to change the library's default error handling behavior.
```
Comment (c24t) [inline specification/error-handling.md]: Removed

> There are situations when end-user wants to know whether API/SDK are used
> correctly. For instance, it may be desirable to not deploy an app with the
> malformed monitoring configuration.

since we say we're allowed to fail fast on configuration errors above.

File: specification/error-handling.md (hunk 8 above)
Comment (carlosalberto) [reply specification/error-handling.md]: To me it *sounds* like it was an example of an Exception, instead of that actual `OutOfMemoryException` (then again, it's probably redundant and we can get rid of that, after your recent changes :) )

File: specification/error-handling.md (hunk 7 above)
Comment (arminru) [reply specification/error-handling.md]: For me this contradicts the statement above:
> We assume that users would prefer to lose telemetry data rather than have the library significantly change the behavior of the instrumented application.

Do we really want to crash the application on startup if there's an issue with the configuration or if the monitoring backend can't be reached?
I'd rather provide means for the user to proactively check if OTel is correctly initialized. I think this check is something we can expect application developers to add to their startup routine. There they can add appropriate error handling (e.g. reporting the misconfiguration to their logging service - or crashing their app if they really want to).

Comment (carlosalberto) [reply specification/error-handling.md]: @arminru I think the motivation is that, if something fails at start time, you can easily/quickly check what's going on, opposed to suddenly get errors once the application has been running for a while.

That being said, we can also probably get more relaxed there, and *only* fail for fatal cases (again, only at the start).

File: specification/error-handling.md (hunk 8 above)
Comment (c24t) [reply specification/error-handling.md]: I took it to mean that exporters should be careful about catching unchecked exceptions that they might cause, in this case by causing an OOM because the export queue grows unbounded.

File: specification/error-handling.md (hunk 7 above)
Comment (c24t) [reply specification/error-handling.md]: > Do we really want to crash the application on startup if there's an issue with the configuration or if the monitoring backend can't be reached?

I think yes if we've got a malformed config, no if the backend can't be reached -- and hopefully we're not making any network requests at startup.

**Comment by Oberon00** [comment]
Note that the Baggage API proposed in open-telemetry/oteps#42 will need different considerations, since it does not deal with telemetry data but data that the application logic may depend on.

**Comment by c24t** [comment]
@open-telemetry/specs-approvers I'm looking for approvals or arguments to move this PR forward.

**Comment by mwear** [comment]
Can we clarify what we should do for dynamic languages when invalid types are used in API methods? We really don't know the types of arguments until the API methods are called at runtime. Should we do runtime type checks or should we just assume arguments are of the right type and try to use them? If we take the latter approach, what should we do for API methods with non-void return types when an exception occurs? We can rescue the exception, but what should we return?

File: specification/error-handling.md (hunk 6 above)
Comment (OlivierAlbertini) [inline specification/error-handling.md]: This doc is great! Should we add how those exceptions can effect the span status ?
Use case: As a user, I provide a callback in order to add custom attributes to my spans. This callback throws an exception. However, beside this exception, everything is ok.  In Zipkin/Jaeger, Should I see: 
- this span in error, 
- no exception to the span (CanonicalCode is `OK`)
- CanonicalCode UNKNOWN with error attributes
- error attributes  but CanonicalCode is `OK`
- an event describing the error and CanonicalCode is `OK`

I think that user should know that everything is slowing down his application. Let me know! Thanks in advance.

Comment (Oberon00) [reply specification/error-handling.md]: 👍  I think that such "internal instrumentation error" information on spans that describes errors not in the instrumented code but the instrumentation itself might be very useful but have not been considered at all yet.

Comment (SergeyKanzhelev) [reply specification/error-handling.md]: @OlivierAlbertini SDK will never know the intention of a callback to add an attribute. I think SDK design must try to avoid those kind of uncertainties. Errors can be exposed as language-specific logs.

Comment (SergeyKanzhelev) [reply specification/error-handling.md]: Also https://github.com/open-telemetry/opentelemetry-specification/issues/275

**Comment by SergeyKanzhelev** [comment]
@mwear what would be your suggestion for the dynamic languages? Would app just crash if you will attempt to take non-existing property or it will return `null`?

File: specification/error-handling.md (hunk 6 above)
Comment (OlivierAlbertini) [reply specification/error-handling.md]: > @OlivierAlbertini SDK will never know the intention of a callback to add an attribute. I think SDK design must try to avoid those kind of uncertainties. Errors can be exposed as language-specific logs.

We have this option coming from OpenCensus on the http plugin. Should we remove those options ?

**Comment by mwear** [comment]
@SergeyKanzhelev I see three possibilities for dynamic languages and they all have their pros and cons

* The contract between the SDK and user is that the SDK will not crash if the APIs are used properly
  * This will be the cleanest and most performant solution
* The SDK can do runtime type checks and provide defaults when they fail
  * Runtime type checks are not free, and will impact performance
  * These checks can be error prone, and exceptions could still occur despite best efforts
* The SDK can rescue errors and return nulls when they are encountered
  * This leads to a fairly poor user experience as they will have to check for nulls on each API call

I think in general, the issue at hand is the lack of type safety in dynamic languages. In general, the responsibility of using an API correctly is passed to the user in these languages. I think it would be reasonable to consider doing the same in OpenTelemetry.

**Comment by SergeyKanzhelev** [comment]
This document calls into the behavior 3 with the exception to return no-op or empty implementations when errors happen.

> The SDK can rescue errors and return nulls when they are encountered
> This leads to a fairly poor user experience as they will have to check for nulls on each API call

**Comment by SergeyKanzhelev** [comment]
@mwear added this paragraph:

> ## Error handling and performance
> 
> Error handling and extensive input validation may cause performance degradation,
> especially on dynamic languages where the input object types are not guaranteed
> in compile time. Runtime type checks will impact performance and are error
> prone, exceptions may occur despite the best effort.
> 
> It is recommended to have a global exception handling logic that will guarantee
> that exceptions are not leaking to the user code. And make a reasonable trade
> off of the SDK performance and fullness of type checks that will provide a
> better on-error behavior and SDK errors troubleshooting.

**Comment by mwear** [comment]
@SergeyKanzhelev: I like Error Handing and Performance paragraph you added above. I think the approach you describe in https://github.com/open-telemetry/opentelemetry-specification/pull/153#issuecomment-537131020 makes sense, although, I don't see that explicitly mentioned in the document. Would it make sense to be more explicit, or are we leaving that out to give more freedom in implementation.

**Comment by SergeyKanzhelev** [comment]
@mwear I've added a note into guidance on returning no-op instead of `null`

**Comment by mwear** [comment]
Thanks @SergeyKanzhelev I'll bring this up at the Ruby SIG so folks are aware of it.

**Comment by c24t** [comment]
> * The SDK can do runtime type checks and provide defaults when they fail
>   
>   * Runtime type checks are not free, and will impact performance
>   * These checks can be error prone, and exceptions could still occur despite best efforts
> * The SDK can rescue errors and return nulls when they are encountered
>   
>   * This leads to a fairly poor user experience as they will have to check for nulls on each API call

I think there's a fourth option here: that the SDK can rescue errors and return non-null defaults when they're encountered. If it's possible to provide defaults for every argument ("do runtime type checks and provide defaults when they fail") then presumably it's possible to provide the default return value that you'd get by calling the function with all the default args.

I don't mean to say this is the right solution, just that it's worth considering with the others.

File: specification/error-handling.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+to the user, the library SHOULD log the error using language-specific
+conventions. SDKs MAY expose callbacks to allow end users to handle
+self-diagnostics separately from application code.
+
+## Exceptions to the rule
+
+SDK authors MAY supply settings that allow end users to change the library's
+default error handling behavior. Application developers may want to run with
+strict error handling in a staging environment to catch invalid uses of the API,
+or malformed config.
```
Comment (c24t) [inline specification/error-handling.md]: @bogdandrutu you brought up the auditing use case in the SIG call, what do you think about expanding on strict error handling here? E.g.:

Under certain circumstances, it's preferable for an application to fail to complete an operation than to complete it but fail to produce telemetry data.

SDK authors SHOULD provide a configuration option that allows the end user to change the library's default error handling behavior. When this option is enabled, the SDK MUST NOT suppress errors or return placeholder objects.

Application developers may also want to run with strict error handling during tests, or in a staging environment, to catch invalid uses of the API, or malformed config.

Comment (c24t) [reply specification/error-handling.md]: @reyang IIRC you're interested in this too.

**Comment by mwear** [comment]
> I think there's a fourth option here: that the SDK can rescue errors and return non-null defaults when they're encountered. If it's possible to provide defaults for every argument ("do runtime type checks and provide defaults when they fail") then presumably it's possible to provide the default return value that you'd get by calling the function with all the default args.
> 
> I don't mean to say this is the right solution, just that it's worth considering with the others.

@c24t Is your comment the same as what is being suggested here: https://github.com/open-telemetry/opentelemetry-specification/pull/153#issuecomment-537131020. If so, that is the approach that this document currently recommends.

File: specification/error-handling.md (hunk 13 above)
Comment (SergeyKanzhelev) [reply specification/error-handling.md]: If we want to standardize this I'd suggest to have a separate work stream on it. "MAY" can be enough for a first iteration

**Comment by c24t** [comment]
> If so, that is the approach that this document currently recommends.

Yup. We're on the same page, https://github.com/open-telemetry/opentelemetry-specification/pull/153#issuecomment-537090903 seemed to suggest those were the only three options.

**Comment by c24t** [comment]
a8451f1 removes hard line breaks to match #192. It doesn't change the text otherwise.

It's a bikesheddy change, I know, but I think it's better to bite the bullet now than keep submitting PRs with text in two different formats.

**Comment by SergeyKanzhelev** [comment]
@iredelmeier, thank you for review. I feel that it is addressed now and ask to review again. 

This proposal was discussed on today's spec SIG and a few before this. And there is a general feeling that we need to document error handling principles.

> The doc currently mingles API and SDK error handling in a way that I expect to be confusing to multiple parties, particularly anyone either working on or using an alternative to the SDK.
>
> Part of the problem may be upstream, i.e., that the API spec is missing logic and so the SDK is forced to backfill (and therefore error handle) it.

This doc asks all implementations to be consistent in their error handling practices: "OpenTelemetry implementations MUST NOT throw unhandled exceptions at run time.". Even though it may be hard to control, this is what we believe must simplify operation with the OpenTelemetry given it's nature of a monitoring, not a business critical tool in most cases.

> While very valuable, self-diagnostics are inherently complicated. This feels like something that needs a larger discussion.

What forum do you suggest this discussion to be held on?

File: specification/error-handling.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+This makes for some unique requirements with respect to error handling.
+
+## Basic error handling principles
+
+OpenTelemetry implementations MUST NOT throw unhandled exceptions at run time.
+
+1. API methods MUST NOT throw unhandled exceptions when used incorrectly by end users.
+   The API and SDK SHOULD provide safe defaults for missing or invalid arguments.
+   For instance, a name like `empty` may be used if the user passes in `null` as the span name argument during `Span` construction.
+2. The API or SDK may _fail fast_ and cause the application to fail on initialization, e.g. because of bad user config or a environment, but MUST NOT cause the application to fail at run time, e.g. due to dynamic config settings received from the Collector.
```
Comment (arminru) [inline specification/error-handling.md]:
```suggestion
2. The API or SDK may _fail fast_ and cause the application to fail on initialization, e.g. because of a bad user config or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.
```
Initialization is also done at run time. It was meant to prohibit crashing _after_ initialization, right?

**Comment by SergeyKanzhelev** [comment]
this PR had a lot of bake time after was approved. Merging
